### DISCUSSION TITLE
how to make a 'collection'?

### DISCUSSION THREAD

**Discussion by majidaldo** [pythological/kanren#73] [General]
Here's a super simple example problem that I run into but similar situations show up many times. Given a list of elements, create all subsets into a variable (to be used elsewhere) but *not* as a set of solutions. Other examples: permutations & combinations, products (stuff you might see in itertools).

Your typical minikanren examples solve these in a 'flat' way whereas I'm interested in 1-to-many relations. Or I may want to relatate particular elements in a list (so I need the full list evaluation).

Another way to think about this: if I'm able to query out what I expect, how do I collect those solutions and put them in a relation?

I don't know how to handle this! (Hope I've been able to communicate this problem).

**Comment by brandonwillard** [comment]
I think I'll need a MWE/pseudo-code in order to understand the situation(s) sufficiently.

**Reply by majidaldo** [reply]
here's a product example
```
from kanren import var, run
from kanren.goals import eq, membero, itero

l = list("abc")
print(
    *run(
        0,
        var("p"),
        membero(var("a"), l),
        membero(var("b"), l),
        eq(var("p"), (var("a"), var("b"))),
        # membero(var('p'), var('plist')   ),  # how to get the list as 1 solution?
    ),
    sep="\n"
)
```
gives:
```
('a', 'a')
('b', 'a')
('c', 'a')
('a', 'b')
('b', 'b')
('c', 'b')
('a', 'c')
('b', 'c')
('c', 'c')
```
...I'm looking to 'collect' this list and use it as *one* data structure to be used in some other relation.

**Reply by brandonwillard** [reply]
OK, I see now.  

Well, in general, I think you would need to reformulate it so that the permutations are generated recursively, because then you could `conso` each result.  More specifically, there might be a way to do this using `kanren.graph.mapo` (or one of the other goals in that module).
